File: benchmarks/modelBenchmark.jl
#=
using Models.TRNmodel
using OrdinaryDiffEq
using MAT


function main()
startTime  = 0.0
endTime    = 1000.0
tspan = (startTime, endTime)
numNeurons = [collect(2:2:18); 100]

reps = 100

tlocal = zeros(length(numNeurons))

for n in eachindex(numNeurons)
    namesOfNeurons = ["TRN$i" for i in 1:numNeurons[n]]

    ## Initial conditions
    u0, per_neuron = initialconditions(numNeurons[n],false)

    ## Initialize simParams
    Params = Vector{simParams}(undef,reps)

    for i = 1:reps
        p = simParams(names=namesOfNeurons,n=numNeurons[n],per_neuron=per_neuron)

        # Vars
        ## Inputs
        for ii = 1:p.n
            p.iDC[ii] = 1.0
            p.iStart[ii] = [200.0]
            p.iStop[ii]  = [800.0]
        end

        Params[i] = p
    end

    tlocal[n] = @elapsed Threads.@threads for i = 1:reps
        p = Params[i]

        prob = ODEProblem(dsim!, u0, tspan, p)

        # Start sim
        solve(prob,BS3(),save_everystep=false,save_idxs=1:p.per_neuron:length(u0))
    end
end

matwrite(joinpath(pwd(),"texec_model_nonoise.mat"),Dict("tlocal"=>tlocal))

=#

using Models.TC_TRNnetwork
using OrdinaryDiffEq
using MAT

function main()
startTime  = 0.0
endTime    = 1000.0
tspan = (startTime, endTime)
numNeurons = [collect(2:2:18); 100]

reps = 100

tlocal = zeros(length(numNeurons))

for n in eachindex(numNeurons)
    namesOfNeurons = [["TRN$i" for i in 1:div(numNeurons[n],2)]; ["TC$i" for i in 1:div(numNeurons[n],2)]]
    ## Initial conditions
    u0, per_neuron = initialconditions(numNeurons[n])

    ## Initialize simParams
    Params = Vector{simParams}(undef,reps)

    for i = 1:reps
        p = simParams(names=namesOfNeurons,n=numNeurons[n],per_neuron=per_neuron)

        # Vars
        ## Inputs
        for ii = 1:p.n
            p.bias[ii] = 0.3
            # noise
            p.A[ii] = 0.5
            p.tA[ii] = poissonP(80,endTime)
            p.AI[ii] = 0.5
            p.tAI[ii] = poissonP(20,endTime)
        end

        Params[i] = p
    end

    tlocal[n] = @elapsed Threads.@threads for i = 1:reps
        p = Params[i]

        prob = ODEProblem(dsim!, u0, tspan, p)

        # Start sim
        solve(prob,BS3(),save_everystep=false,save_idxs=1:p.per_neuron:length(u0))
    end
end

matwrite(joinpath(pwd(),"texec_TC_TRNnet_noise.mat"),Dict("tlocal"=>tlocal))




tlocal = zeros(length(numNeurons))

for n in eachindex(numNeurons)
    namesOfNeurons = [["TRN$i" for i in 1:div(numNeurons[n],2)]; ["TC$i" for i in 1:div(numNeurons[n],2)]]

    ## Initial conditions
    u0, per_neuron = initialconditions(numNeurons[n],false)

    ## Initialize simParams
    Params = Vector{simParams}(undef,reps)

    for i = 1:reps
        p = simParams(names=namesOfNeurons,n=numNeurons[n],per_neuron=per_neuron)

        # Vars
        ## Inputs
        for ii = 1:p.n
            p.iDC[ii] = 1.0
            p.iStart[ii] = [200.0]
            p.iStop[ii]  = [800.0]
        end

        Params[i] = p
    end

    tlocal[n] = @elapsed Threads.@threads for i = 1:reps
        p = Params[i]

        prob = ODEProblem(dsim!, u0, tspan, p)

        # Start sim
        solve(prob,BS3(),save_everystep=false,save_idxs=1:p.per_neuron:length(u0))
    end
end

matwrite(joinpath(pwd(),"texec_TC_TRNnet_nonoise.mat"),Dict("tlocal"=>tlocal))


nothing
end #main

main()

File: benchmarks/modelBenchmark.m
%Simulation//Run variables
namesOfNeurons = {'TRN1','TRN2','TRN3','TRN4','TRN5','TRN6','TRN7','TRN8','TRN9',...
                  'TC1', 'TC2', 'TC3', 'TC4', 'TC5', 'TC6', 'TC7', 'TC8', 'TC9' };
num_pairs = 9;

startTime  = 0;
endTime    = 1000;
tspan = [startTime endTime];
options=odeset('InitialStep',10^(-2),'MaxStep',10^(-1));
reps = 100;
numWorkers = [16 8];
tLocal = zeros(3,9);

for w = 1:numel(numWorkers)
% Connecting parallel pool
c = parcluster;
parpool(c, numWorkers(w));


% TRN network
for n = 1:num_pairs
n_names = {namesOfNeurons{[1:n,10:9+n]}};
load('s_init.mat','s_init')
s_init = [s_init zeros(1,length(n_names)/2)];
s0 = repmat(s_init,1,length(n_names));
per_neuron = length(s_init);

    tic;
    parfor i = 1:reps
        sim = simParams(n_names,per_neuron,s0);
        for ii=1:sim.n
        sim.iDC(ii) = 0.3;
        sim.istart{ii} = 0;
        sim.istop{ii} = 1000;
        %
        sim.A(ii) = 0.2;
        sim.tA{ii} = poissonP(80,endTime);
        sim.AI(ii) = 0.2;
        sim.tAI{ii} = poissonP(20,endTime);
        end
        
        [t,s] = ode23(@(t,s) dsim(t,s,sim),tspan,sim.s0,options);
    end
    tLocal(w,n) = toc;
end
delete(gcp('nocreate')) %shutdown parallel pool
end


% Connecting parallel pool
parpool('Threads')

% TRN network w/ noise
for n = 1:num_pairs
n_names = {namesOfNeurons{[1:n,10:9+n]}};
load('s_init.mat','s_init')
s_init = [s_init zeros(1,length(n_names)/2)];
s0 = repmat(s_init,1,length(n_names));
per_neuron = length(s_init);

    tic;
    parfor i = 1:reps
        sim = simParams(n_names,per_neuron,s0);
        for ii=1:sim.n
        sim.iDC(ii) = 0.3;
        sim.istart{ii} = 0;
        sim.istop{ii} = 1000;
        %
        sim.A(ii) = 0.2;
        sim.tA{ii} = poissonP(80,endTime);
        sim.AI(ii) = 0.2;
        sim.tAI{ii} = poissonP(20,endTime);
        end
        
        [t,s] = ode23(@(t,s) dsim(t,s,sim),tspan,sim.s0,options);
    end
    tLocal(3,n) = toc;
end
delete(gcp('nocreate')) %shutdown parallel pool

save("exec_time_matlab.mat","tLocal")
